### PR TITLE
Add mypy types to visa backend

### PR DESCRIPTION
**Pull request by jenshnielsen**
Noticed from #1738 that this important code is not typed

### CODE REVIEW

File: qcodes/instrument/visa.py [hunk 1]
```diff
@@ -24,37 +24,36 @@ class VisaInstrument(Instrument):
...
-
+        name: What this instrument is called locally.
+        address: The visa resource name to use to connect.
+        timeout: seconds to allow for responses. Default 5.
         terminator: Read termination character(s) to look for. Default ``''``.
-
         device_clear: Perform a device clear. Default True.
-
+        visalib: Visa backend to use when connecting to this instrument.
+            This should be in the form of a string '@<backend>'.
```
Comment (jenshnielsen) [inline qcodes/instrument/visa.py]: Why is this like that you may ask. I think we should remove this logic and just make it a string in a different pr

File: qcodes/instrument/visa.py [hunk 2]
```diff
@@ -159,17 +156,17 @@ def set_terminator(self, terminator: str):
...
+    def _set_visa_timeout(self, timeout: Optional[Union[float, int]]):
 
         if timeout is None:
             self.visa_handle.timeout = None
         else:
             # pyvisa uses milliseconds but we use seconds
             self.visa_handle.timeout = timeout * 1000.0
 
-    def _get_visa_timeout(self):
+    def _get_visa_timeout(self) -> None:
```
Comment (astafan8) [inline qcodes/instrument/visa.py]:
```suggestion
    def _get_visa_timeout(self) -> Optional[float]:
```

File: qcodes/instrument_drivers/rigol/DG1062.py [hunk 3]
```diff
@@ -344,7 +345,7 @@ class DG1062(VisaInstrument):
     waveforms = DG1062Channel.waveforms
 
     def __init__(self, name: str, address: str,
-                 **kwargs: Dict) ->None:
+                 **kwargs) -> None:
```
Comment (astafan8) [inline qcodes/instrument_drivers/rigol/DG1062.py]: no need for it as it is implicit
```suggestion
                 **kwargs):
```

File: qcodes/instrument/visa.py [hunk 4]
```diff
@@ -24,37 +24,36 @@ class VisaInstrument(Instrument):
...
-
         device_clear: Perform a device clear. Default True.
-
+        visalib: Visa backend to use when connecting to this instrument.
+            This should be in the form of a string '@<backend>'.
+            By default the NI backend is used, but '@py' will use the
+            ``pyvisa-py`` backend. Note that QCoDeS does not install (or even require)
+            ANY backends, it is up to the user to do that. see eg:
+            http://pyvisa.readthedocs.org/en/stable/names.html
         metadata (Optional[Dict]): additional static metadata to add to this
```
Comment (jenshnielsen) [inline qcodes/instrument/visa.py]: This is not a actually an explicit kwarg here and it is unused in the base class (`Instrument`). We should either remove it or actually use it

File: qcodes/instrument/visa.py (hunk 1 above)
Comment (astafan8) [reply qcodes/instrument/visa.py]: do you mean the `@` logic? i believe that it comes from pyvisa

Comment (jenshnielsen) [reply qcodes/instrument/visa.py]: Yes but that does not mean that it's not pointless :) As far as I can see the only thing `self.visalib` is used for is to split off the `@` to form `self.visabackend` in `set_adress`

File: qcodes/instrument/visa.py (hunk 4 above)
Comment (Dominik-Vogel) [reply qcodes/instrument/visa.py]: That is interesting. I think it should have gone into `Metadatable` but somebody forgot to put it into the `super().__init__` call in `InstrumentBase`.

File: qcodes/instrument/visa.py [hunk 5]
```diff
@@ -24,37 +24,36 @@ class VisaInstrument(Instrument):
...
+    See help for :class:`qcodes.Instrument` for additional information on writing
     instrument subclasses.
 
     Attributes:
         visa_handle (pyvisa.resources.Resource): The communication channel.
     """
 
-    def __init__(self, name, address=None, timeout=5,
-                 terminator='', device_clear=True, visalib=None, **kwargs):
+    def __init__(self, name: str, address: str, timeout: Union[int, float] = 5,
```
Comment (jenshnielsen) [inline qcodes/instrument/visa.py]: I do not see the point of the default value of address being None. This however does break a test of simulated instruments

File: qcodes/instrument/visa.py (hunk 4 above)
Comment (jenshnielsen) [reply qcodes/instrument/visa.py]: This will be resolved in #1740

File: qcodes/instrument/visa.py (hunk 5 above)
Comment (astafan8) [reply qcodes/instrument/visa.py]: that's good, i agree.

File: qcodes/instrument/visa.py (hunk 1 above)
Comment (astafan8) [reply qcodes/instrument/visa.py]: ok, understood. yes, i tend to agree. and then perhaps we can finally remove the logic about warnings and `@` being in `address` and `visalib` arguments.

File: qcodes/utils/validators.py [hunk 6]
```diff
@@ -331,7 +331,7 @@ class Enum(Validator):
     eg. Enum(val1, val2, val3)
     """
 
-    def __init__(self, *values: Hashable) -> None:
+    def __init__(self, *values: Union[Hashable, None]) -> None:
```
Comment (astafan8) [inline qcodes/utils/validators.py]:
```suggestion
    def __init__(self, *values: Optional[Hashable]) -> None:
```

File: qcodes/tests/test_visa.py [hunk 7]
```diff
@@ -91,7 +92,7 @@ class TestVisaInstrument(TestCase):
     ]
 
     def test_ask_write_local(self):
-        mv = MockVisa('Joe')
+        mv = MockVisa('Joe', 'none')
```
Comment (astafan8) [inline qcodes/tests/test_visa.py]: for readability?
```suggestion
        mv = MockVisa('Joe', 'none_address')
```
